260 user messages

snasif opened the pull request.

snasif committed:
Fixed some issues with remove_subscriber & add_subscriber

Update to admin authority to add super user and fixed code to prevent sender from removing themselves.

snasif committed:
added generate_stats to use /stats to get message count for user in a time period

snasif committed:
continued working on /stats command, need testing.

snasif committed:
Update .gitignore

snasif committed:
Task 258 completed

snasif committed:
Task 259 completed

snasif committed:
minor update

hkcountryman committed:
Merge branch 'dev' into 260-user-messages

Notes in general for everyone's PRs:

- If a function is "private" preface its name with an underscore. Then
  you'll get a linter warning if attempting to use it outside the module
  it was intended for.
- If you're returning an error or success message in English, write
  support to translate it in `language_data.py` or reuse an existing
  error from there.
- In addition to function docs, update the docs at the top of each
  module. All "public" functions (the ones that don't start with
  underscores) and all classes should be summarized up there. And of
  course classes also have their own docstrings; see `Chatbot` class if
  you need an example.

TODO items we still need to fix: these do not necessarily pertain to
this PR but we should finish them this week...

- Would it make sense to keep all JSON files under the `json/`
  directory? And probably rename `team56test.json` so it is clearer what
  it's for. Make sure to do this with find and replace in the entire
  project so that all non-code references to such a file get changed if
  we go this route.
- Create log file in `setup.py`, and possibly a backup for it.
- Log file stuff was also added to the constructor for `Chatbot`, so do
  we need to do encryption and backup for this too? And would we need to
  also deal with encryption in `_store_message_timestamp`? (Kevin!)
- Better error handling in constructor for `Chatbot` when a file is
  corrupted
- Make username and phone number interchangeable identifiers for the
  following functions in `chatbot.py`: `_query`, '_remove_subscriber`,
  and `_get_last_post_time`
- In `_add_subscriber`, make sure to also add an empty entry for the new
  user in the timestamp logs. In `_remove_subscriber`, make sure to also
  remove the subscriber's timestamped chat logs.
- Use translated error and success messages in the following functions:
  `_generate_stats`, `_get_last_post_time`, and `_generate_total_stats`
- Make `/list` have tabular output, refactor `/stats` and `/totalstats`
  to generate a tabular batch report (`/stats n day`), and make
  `/lastpost` with no args return a tabular batch report for all users.

Other: other stuff from the sponsor meeting I'm noting again here for my
own sanity but they do not have associated TODO comments...

- Change the format of the timestamp log file so that it is more like
  `{ "user": [{"date": # posts}, ...], ...}`
- Only keep the last year of data in the logs
- Always check documentation is up to date
- Come up with tests

hkcountryman reviewed (commented):
**Notes in general for everyone's PRs:**

- If a function is "private" preface its name with an underscore. Then you'll get a linter warning if attempting to use it outside the module it was intended for.
- If you're returning an error or success message in English, write support to translate it in `language_data.py` or reuse an existing error from there.
- In addition to function docs, update the docs at the top of each module. All "public" functions (the ones that don't start with underscores) and all classes should be summarized up there. And of course classes also have their own docstrings; see `Chatbot` class if you need an example.

**TODO items we still need to fix**: these do not necessarily pertain to this PR but we should finish them this week...

- Would it make sense to keep all JSON files under the `json/` directory? And probably rename `team56test.json` so it is clearer what it's for. Make sure to do this with find and replace in the entire project so that all non-code references to such a file get changed if we go this route.
- Create log file in `setup.py`, and possibly a backup for it.
- Log file stuff was also added to the constructor for `Chatbot`, so do we need to do encryption and backup for this too? And would we need to also deal with encryption in `_store_message_timestamp`? @kyang72 
- Better error handling in constructor for `Chatbot` when a file is corrupted
- Make username and phone number interchangeable identifiers for the following functions in `chatbot.py`: `_query`, `_remove_subscriber`, and `_get_last_post_time`
- In `_add_subscriber`, make sure to also add an empty entry for the new user in the timestamp logs. In `_remove_subscriber`, make sure to also remove the subscriber's timestamped chat logs.
- Use translated error and success messages in the following functions: `_generate_stats`, `_get_last_post_time`, and `_generate_total_stats`
- Make `/list` have tabular output, refactor `/stats` and `/totalstats` to generate a tabular batch report (`/stats n day`), and make `/lastpost` with no args return a tabular batch report for all users.

**Other:** other stuff from the sponsor meeting I'm noting again here for my own sanity but they do not have associated TODO comments...

- Change the format of the timestamp log file so that it is more like `{ "user": [{"date": # posts}, ...], ...}`
- Only keep the last year of data in the logs
- Always check documentation is up to date
- Come up with tests

hkcountryman committed:
Update documentation in Chatbot constructor

Another thing - it occurs to me that it's silly to make users do
    `/stats n day`
because we only offer "day" units. We should either offer more (in which
case I would probably say we should use single-letter identifiers like
d, m, and y) or we should not require that argument.

hkcountryman reviewed (commented):
Another thing - it occurs to me that it's silly to make users do
        `/stats n day`
because we only offer "day" units. We should either offer more (in which case I would probably say we should use single-letter identifiers like d, m, and y) or we should not require that argument.

hkcountryman commented:
@snasif Kevin needed this merged so you'll need to do a new PR if you're still on this branch
